src/filter.jl

```diff
@@ -85,3 +85,206 @@ function hamilton_filter(y::AbstractVector, h::Integer)
     y_trend = y - y_cycle
     return y_cycle, y_trend
 end
+
+"""
+- `g::Function`: Parameterized `Function` that recieves three arguments 
+                 (data at each period, state at each period, parameter) and return the density.
+                 For example, if the data `x <: Real` is generated by `N(mu_s, sigma_s)` where mean `mu` and std `sigma` is
+                 state dependent, `g` should be defined as 
+                 ```
+                 g(x, s, parameter) = exp(-((x-parameter.mu[s])^2)/(2*parameter.sigma[s]^2))/sqrt(2*pi*parameter.sigma[s]^2)
+                 ```
+- `y::AbstractArray`: Collection of data whose row corresponds to periods
+- `parameter`: Collection of parameters in the model. Each parameter must be a vector that stores the values for each state.
+               For example, if you want to set `mu` at state 1 is 0 and `mu` at state 2 is 0.5 and 
+               `sigma` at state 1 is 3 and `sigma` at state 2 is 4, `parameter` should be
+               ```
+               parameter = (mu = [0, 0.5], sigma = [3, 4])
+               ```
+               Alternatively, you can define your own type and use it:
+               ```
+               struct ParaType
+                   mu::Vector
+                   sigma::Vector
+               end
+               parameter = ParaType([0, 0.5], [3, 4])
+               ```
+- `P::AbstractMatrix`: Transition matrix of state. It must be square.
+- `M::Integer`: Number of states. It must be same as the number of rows and columns of `P`. If `M` is not passed, 
+                number of rows of `P` is set as `M`.
+"""
+mutable struct RegimeSwitchingModel{TD <: AbstractArray, TP <: AbstractMatrix, TPS <: AbstractMatrix}
+    g::Function
+    y::TD
+    parameter
+    P::TP
+    M::Int
+    prob_smoothed::TPS
+    logL::Float64
+    function RegimeSwitchingModel(g, y, parameter, P, M, prob_smoothed, logL)
+        if size(P, 1) != size(P, 2)
+            error("P must be square")
+        end
+        if size(P, 1) != M
+            error("the number of rows and columns of P must be equal to M")
+        end
+        return new{typeof(y), typeof(P), typeof(prob_smoothed)}(g, y, parameter, P, M, prob_smoothed, logL)
+    end
+end
+RegimeSwitchingModel(g::Function, y::AbstractArray, parameter, P::AbstractMatrix) =
+    RegimeSwitchingModel(g, y, parameter, P, size(P, 1), fill(NaN, size(y, 1), size(P, 1)), NaN)
+
+"""
+Apply the filter developed by Hamilton (1989, Econometrica) to a regime switching model.
+
+##### Arguments
+- `rsm::RegimeSwitchingModel`: `RegimeSwitchingModel`
+- `prob_update_pre::AbstractArray`: Probability distribution of state at period 0 conditional on the information up to
+                                   period 0.
+
+##### Returns
+- `p`: likelihood at each period

```

Comment:
It wouldn't hurt to emphasize that the returns are each of size `T x M` where `T` is the number of observations and `M` is the number of regimes.

Reply:
That's fair. I'll do so.